app/src/main/java/org/kiwix/kiwixmobile/webserver/WebServerHelper.kt

```diff
@@ -0,0 +1,103 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.webserver
+
+import android.util.Log
+import org.kiwix.kiwixmobile.core.utils.ServerUtils.getIpAddress
+import org.kiwix.kiwixmobile.core.utils.ServerUtils.getIp
+import org.kiwix.kiwixmobile.webserver.wifi_hotspot.IpAddressCallbacks
+import io.reactivex.Flowable
+import io.reactivex.android.schedulers.AndroidSchedulers
+import org.kiwix.kiwixmobile.core.utils.ServerUtils
+import org.kiwix.kiwixmobile.core.utils.ServerUtils.INVALID_IP
+import java.util.ArrayList
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+/**
+ * WebServerHelper class is used to set up the suitable environment i.e. getting the
+ * ip address and port no. before starting the WebServer
+ * Created by Adeel Zafar on 18/07/2019.
+ */
+class WebServerHelper @Inject constructor(
+  private val kiwixServerFactory: KiwixServer.Factory,
+  private val ipAddressCallbacks: IpAddressCallbacks
+) {
+  private var kiwixServer: KiwixServer? = null
+  private var isServerStarted = false
+
+  @Suppress("ReturnCount")
+  fun startServerHelper(selectedBooksPath: ArrayList<String>): Boolean {
+    val ip = getIpAddress()
+    if (ip.isNullOrEmpty()) {
+      return false
+    } else if (startAndroidWebServer(selectedBooksPath)) {
+      return true
+    }
+    return isServerStarted
+  }
+
+  fun stopAndroidWebServer() {
+    if (isServerStarted) {
+      kiwixServer?.stopServer()
+      updateServerState(false)
+    }
+  }
+
+  @Suppress("MagicNumber")
+  private fun startAndroidWebServer(selectedBooksPath: ArrayList<String>): Boolean {
+    if (!isServerStarted) {
+      ServerUtils.port = 8080
+      kiwixServer = kiwixServerFactory.createKiwixServer(selectedBooksPath).also {
+        updateServerState(it.startServer(ServerUtils.port))
+        Log.d(TAG, "Server status$isServerStarted")
+      }
+    }
+    return isServerStarted
+  }
+
+  private fun updateServerState(isStarted: Boolean) {
+    isServerStarted = isStarted
+    ServerUtils.isServerStarted = isStarted
+  }
+
+  // Keeps checking if hotspot has been turned using the ip address with an interval of 1 sec
+  // If no ip is found after 15 seconds, dismisses the progress dialog
+  @Suppress("MagicNumber")
+  fun pollForValidIpAddress() {
+    Flowable.interval(1, TimeUnit.SECONDS)
+      .map { _: Long? -> getIp() }
+      .filter { s: String? -> s != INVALID_IP }
+      .timeout(15, TimeUnit.SECONDS)
+      .take(1)
+      .observeOn(AndroidSchedulers.mainThread())
+      .subscribe(
+        { s: String? ->
+          ipAddressCallbacks.onIpAddressValid()
+          Log.d(TAG, "onSuccess:  $s")
+        }
+      ) { e: Throwable? ->
```

Comment:
We can get rid of the redundant lambda arrows.

Reply:
Done